File: src/modules/pgps.py
#!/usr/bin/python

import socket
import sys
from thread import *
import time
import json
import threading

def generate_error(message):
    print '\033[31m\033[1m%s\033[0m\033[39m' % message
    
try:
    from gps import *
except ImportError:
    generate_error('GPS not configured properly!')
    sys.exit(1)

gpsd = None  # Global GPSD variable
gpsd_data = {}

# Define the socket parameters
HOST = ''
PORT = 8907

connection = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
connection.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1) 

# Bind socket to local host and port
try:
    connection.bind((HOST, PORT))
except socket.error, msg:
    generate_error('Bind failed. Error Code : ' + str(msg[0]) + ' Message ' \
        + msg[1])
    sys.exit()

print 'GPS socket bind complete!'

# GPSD polling class definitions
class GPSPoller(threading.Thread):

    def __init__(self):
        threading.Thread.__init__(self)
        global gpsd  # Bring it in scope
        gpsd = gps(mode=WATCH_ENABLE)  # Starting the stream of info
        self.current_value = None
        self.running = True  # Setting the thread running to true

    def run(self):
        global gpsd
        while gpsp.running:
            gpsd.next()  # This will continue to loop and grab each set of GPSD info to clear the buffer

# Create and start the GPSD thread
gpsp = GPSPoller()
gpsp.daemon = True # Needed to make the thread shutdown correctly
gpsp.start()

# Start listening on socket
connection.listen(10)

# Function for handling connections; will be used to create threads
def clientthread(conn):

    # Infinite loop so that function do not terminate and thread do not end
    while True:

        # Updates the data dictionary        
        update_gpsd_data()

        # Receive data from the client
        data = conn.recv(1024)
        if not data:
            break

        conn.sendall(json.dumps(gpsd_data).encode('utf-8'))

    # Close the connection if the client if the client and server connection is interfered
    conn.close()

def update_gpsd_data():
    try:
        # Tries to pull in all the critical GPS data
        gpsd_data.clear()
        gpsd_data.update({
            'latitude': gpsd.fix.latitude,
            'longitude': gpsd.fix.longitude,
            'timestamp': gpsd.fix.time,
            'heading': gpsd.attitude.heading,
            'speed': gpsd.fix.speed,
            'roll': gpsd.attitude.roll,
            'pitch': gpsd.attitude.pitch,
            'yaw': gpsd.attitude.yaw,
            })
    except AttributeError:
        # Uses standard GPS data for units that have unsupported features
        gpsd_data.update({
            'latitude': gpsd.fix.latitude,
            'longitude': gpsd.fix.longitude,
            'timestamp': gpsd.fix.time,
            'speed': gpsd.fix.speed,
            })

# Main loop to keep the server process going
while True:

    try:
        # Wait to accept a connection in a blocking call
        (conn, addr) = connection.accept()
        print 'Connected with ' + addr[0] + ':' + str(addr[1])

        # Start new thread takes 1st argument as a function name to be run, second is the tuple of arguments to the function
        start_new_thread(clientthread, (conn, ))

    except KeyboardInterrupt, socket.error:
        connection.shutdown(socket.SHUT_RDWR)
        connection.close()
        break

            

File: src/modules/pgps.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <unistd.h>

// Socket specific libraries
#include <sys/types.h> 
#include <sys/socket.h>
#include <netinet/in.h>

// GPSD specific libraries
#include <gps.h>
#include "../lib/gpsd_config.h"
#include "../lib/gpsdclient.h"

static time_t status_timer;
static struct fixsource_t source;
static struct gps_data_t gpsdata;

char json[1024];

// Method prototypes
void build_json(void);
void write_element(char*, float, char*);
void get_data(void);

// GPS variables
int count = 0, track = 0, status = 0, heading = 0;
double latitude = 0, longitude = 0, unix_time = 0;
double roll = 0, pitch = 0, yaw = 0, speed = 0;

// Define error handling methods
void error(const char *msg) {
    perror(msg);
    exit(1);
}

int main(int argc, char *argv[]) {
    printf("[GPS Socket]: Starting the GPS Socket\n");

    // Close GPS in case of prior inproper shutdown
    gps_close(&gpsdata);

    // Flags for GPSD
    unsigned int flags = WATCH_ENABLE;
    char tbuf[CLIENT_DATE_MAX + 1];

    if (optind < argc) {
        gpsd_source_spec(argv[optind], & source);

    } else {
        gpsd_source_spec(NULL, & source);
    }

    // Open the GPSD stream
    if (gps_open("127.0.0.1", "2947", & gpsdata) != 0) {
        error("[GPS Socket]: No GPSD running or network error!\n");
        // Exit with error status
        return 1;
    } else {
        printf("[GPS Socket]: GPS port successfully opened!\n");
    }

    status_timer = time(NULL);

    if (source.device != NULL) {
        flags |= WATCH_DEVICE;
    }

    // Stream GPS Parameters
    (void) gps_stream(&gpsdata, flags, source.device);

    // Setup the socket connection
    int sockfd, clientsockfd, port;

    socklen_t client;

    char buffer[256];
    struct sockaddr_in server_address, client_address;

    int n;

    if (argc < 2) {
        fprintf(stderr, "[GPS Socket]: Error: No port specified\n");
        exit(1);
    }

    // Create the socket connection
    sockfd = socket(AF_INET, SOCK_STREAM, 0);

    if (sockfd < 0) {
        error("[GPS Socket]: Error: Cannot open socket");
    }

    bzero((char*) & server_address, sizeof(server_address));

    port = atoi(argv[1]);
    server_address.sin_family = AF_INET;
    server_address.sin_addr.s_addr = INADDR_ANY;
    server_address.sin_port = htons(port);

    // Bind to the socket
    if (bind(sockfd, (struct sockaddr*) & server_address,
            sizeof(server_address)) < 0) {
        error("[GPS Socket]: Error: Failed to bind to socket");
    }

    listen(sockfd, 5);

    // Create and accept the client socket connection
    client = sizeof(client_address);
    clientsockfd = accept(sockfd, (struct sockaddr*) & client_address, &client);

    if (clientsockfd < 0) {
        error("[GPS Socket]: Error: Problem accepting request");
    }

    // Enter and infinite connection with the client
    for (;;) {
        bzero(buffer, 256);

        // Socket read functions
        n = read(clientsockfd, buffer, 255);

        if (n < 0) {
            error("[GPS Socket]: Error: Cannot read from socket");
        }

        get_data();
        build_json();

        // Socket write functions
        printf("Sent: %s\n", json);
        n = write(clientsockfd, json, 0);


        if (n < 0) {
            error("[GPS Socket]: Error: Cannot write to socket");
        }
    }

    // Close the client and socket connections
    close(clientsockfd);
    close(sockfd);

    return 0;
}

void build_json(void) {
    strcpy(json, "{ ");

    write_element("time", unix_time, "%d");
    write_element("latitude", latitude, "%0.7f");
    write_element("longitude", longitude, "%0.7f");
    write_element("heading", heading, "%f");
    write_element("roll", roll, "%0.3f");
    write_element("pitch", pitch, "%0.3f");
    write_element("yaw", yaw, "%0.3f");

    strcpy(json + strlen(json), " }");
}

void write_element(char* property, float value, char* format) {
    sprintf(json + strlen(json), "\"%s\":", property);
    sprintf(json + strlen(json), format, value);
    strcpy(json + strlen(json), ", ");
}


void get_data(void) {
        if (gps_read(&gpsdata) == -1) {
            // Print error if GPS isn't read
            printf("[GPS Socket]: GPS not read!\n");
        } else {
            if (isnan(gpsdata.fix.latitude) != 0) {
                // Don't use GPS data without a Fix
                printf("[GPS Socket]: No GPS fix!\n");
            }
            else {
                // Get Basic GPS parameters 
                latitude = gpsdata.fix.latitude;
                longitude = gpsdata.fix.longitude;
                unix_time = gpsdata.fix.time;
                track = gpsdata.fix.track;
                speed = gpsdata.fix.speed;

                // Specific to Hemisphere H-102 GPS
                heading = gpsdata.attitude.heading;
                roll = gpsdata.attitude.roll;
                pitch = gpsdata.attitude.pitch;
                yaw = gpsdata.attitude.yaw;
                status = gpsdata.attitude.mag_st;

                if (status == 0) {
                    heading = track;
                }

                if ((isnan(gpsdata.attitude.heading) != 0) || (heading > 360 || heading < 0)) {
                    heading = 0;
                }
                if ((isnan(gpsdata.fix.track) != 0) || (track > 360 || track < 0)) {
                    track = 0;
                }
                if ((isnan(gpsdata.attitude.roll) != 0) || (roll > 360 || roll < 0)) {
                    roll = 0;
                }
                if ((isnan(gpsdata.attitude.pitch) != 0) || (pitch > 360 || pitch < 0)) {
                    pitch = 0;
                }
                if ((isnan(gpsdata.attitude.yaw) != 0) || (yaw > 360 || yaw < 0)) {
                    yaw = 0;
                }
            }
        }
}